File: Interview Preparation Kit/Stacks and Queues/BalancedBrackets.go
// https://www.hackerrank.com/challenges/balanced-brackets/problem
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

/*
 * Complete the 'isBalanced' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts STRING s as parameter.
 */
func isBalanced(s string) string {
	stack := make([]rune, 0)

	for _, bracket := range s {
		if bracket == '(' || bracket == '{' || bracket == '[' {
			stack = append(stack, bracket)
		} else {
			if len(stack) == 0 {
				return "NO"
			}

			last_index := len(stack) - 1
			top := stack[last_index]
			if bracket == ')' && top == '(' {
				stack = stack[:last_index]
			} else if bracket == '}' && top == '{' {
				stack = stack[:last_index]
			} else if bracket == ']' && top == '[' {
				stack = stack[:last_index]
			} else {
				return "NO"
			}
		}
	}

	if len(stack) == 0 {
		return "YES"
	}
	return "NO"
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16*1024*1024)

	tTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
	checkError(err)
	t := int32(tTemp)

	for tItr := 0; tItr < int(t); tItr++ {
		s := readLine(reader)

		result := isBalanced(s)

		fmt.Fprintf(writer, "%s\n", result)
	}

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}

File: Interview Preparation Kit/Stacks and Queues/BalancedBrackets.js
// https://www.hackerrank.com/challenges/balanced-brackets/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

// Complete the isBalanced function below.
function isBalanced(s) {
    let stack = []

    for (let i = 0; i < s.length; i++) {
        let bracket = s[i];
        if (bracket == '(' || bracket == '{' || bracket == '[') {
            stack.push(bracket);
        }
        else {
            if (stack.length == 0) {
                return "NO";
            }

            let top = stack[stack.length - 1];
            if (bracket == ')' && top == '(') {
                stack.pop();
            }
            else if (bracket == '}' && top == '{') {
                stack.pop();
            }
            else if (bracket == ']' && top == '[') {
                stack.pop();
            }
            else {
                return "NO";
            }
        }
    }

    if (stack.length == 0) {
        return "YES";
    }

    return "NO";
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const t = parseInt(readLine(), 10);

    for (let tItr = 0; tItr < t; tItr++) {
        const s = readLine();

        let result = isBalanced(s);

        ws.write(result + "\n");
    }

    ws.end();
}
